File: Python/Day13_8_05_2025/TicTacToe.py
# Q3. TicTacToe program
def print_board(board):
    for row in board:
        print(" | ".join(row))
        print("-" * 5)
def check_winner(board, player):
    for row in board:
        if all(cell == player for cell in row):
            return True
    for col in range(3):
        if all(board[row][col] == player for row in range(3)):
            return True
    if all(board[i][i] == player for i in range(3)):
        return True
    if all(board[i][2 - i] == player for i in range(3)):
        return True
    return False
def is_draw(board):
    return all(cell in ['X', 'O'] for row in board for cell in row)
def get_move(player):
    while True:
        try:
            row = int(input(f"{player}'s Turn - Enter row (0-2): "))
            col = int(input(f"{player}'s Turn - Enter column (0-2): "))
            if row in range(3) and col in range(3):
                return row, col
            else:
                print("Invalid input. Please enter values between 0 and 2.")
        except ValueError:
            print("Invalid input. Please enter numbers only.")
def play_game():
    board = [[" " for _ in range(3)] for _ in range(3)]
    current_player = "X"
    while True:
        print_board(board)
        row, col = get_move(current_player)
        if board[row][col] == " ":
            board[row][col] = current_player
        else:
            print("Cell already taken. Choose another.")
            continue
        if check_winner(board, current_player):
            print_board(board)
            print(f"Player {current_player} wins!")
            break
        elif is_draw(board):
            print_board(board)
            print("It's a draw!")
            break
        current_player = "O" if current_player == "X" else "X"
if __name__ == "__main__":
    play_game()

File: Java/Day7_8_05_2025/src/TicTacToe.java
// Q7. TicTacToe program in java using method.
import java.util.Scanner;
public class TicTacToe {
    static char[][] board = {
        {' ', ' ', ' '},
        {' ', ' ', ' '},
        {' ', ' ', ' '}
    };
    static Scanner scanner = new Scanner(System.in);
    public static void main(String[] args) {
        char currentPlayer = 'X';
        boolean gameEnded = false;
        System.out.println("Welcome to Tic-Tac-Toe!");
        while (!gameEnded) {
            printBoard();
            playerMove(currentPlayer);
            if (checkWin(currentPlayer)) {
                printBoard();
                System.out.println("Player " + currentPlayer + " wins!");
                gameEnded = true;
            } else if (isDraw()) {
                printBoard();
                System.out.println("It's a draw!");
                gameEnded = true;
            } else {
                currentPlayer = (currentPlayer == 'X') ? 'O' : 'X'; // Switch player
            }
        }
    }
    public static void printBoard() {
        System.out.println("Board:");
        for (int i = 0; i < 3; i++) {
            System.out.println(" " + board[i][0] + " | " + board[i][1] + " | " + board[i][2]);
            if (i < 2) System.out.println("---|---|---");
        }
    }
    public static void playerMove(char player) {
        int row, col;
        while (true) {
            System.out.println("Player " + player + ", enter your move (row and column: 0 1 2):");
            row = scanner.nextInt();
            col = scanner.nextInt();
            if (row >= 0 && row < 3 && col >= 0 && col < 3) {
                if (board[row][col] == ' ') {
                    board[row][col] = player;
                    break;
                } else {
                    System.out.println("Cell already taken. Choose again.");
                }
            } else {
                System.out.println("Invalid input. Try again.");
            }
        }
    }
    public static boolean checkWin(char player) {
        for (int i = 0; i < 3; i++) {
            if ((board[i][0] == player &&
                 board[i][1] == player &&
                 board[i][2] == player) ||
                (board[0][i] == player &&
                 board[1][i] == player &&
                 board[2][i] == player)) {
                return true;
            }
        }
        return (board[0][0] == player &&
                board[1][1] == player &&
                board[2][2] == player) ||
                (board[0][2] == player &&
                board[1][1] == player &&
                board[2][0] == player);
    }
    public static boolean isDraw() {
        for (int i = 0; i < 3; i++) {
            for (int j = 0; j < 3; j++) {
                if (board[i][j] == ' ') {
                    return false;
                }
            }
        }
        return true;
    }
}